src/partisan_hiscamp_membership_strategy.erl

```diff
@@ -0,0 +1,251 @@
+%% Membership0 is a set of node elements
+join(#hiScamp{membership=Membership0, level1=L1, level2=L2}=State0, Node, _NodeState) ->
+    OutgoingMessages0 = [],
+    case partisan_config:get(tracing, ?TRACING) of
+        true ->
+            lager:info("~p: Adding node ~p to our membership.", [node90, Node]);
+        false ->
+            ok
+    end,
+    Membership = sets:add_element(Node, Membership0),
+%% notify node to add us to its state.
+    Myself = partisan_peer_service_manager:myself(),
+    OutgoingMessages1 = OutgoingMessages0 ++ [{Node, {protocol, {forward_subscription, Myself}}}],
+    OutgoingMessages2 = sets:fold(fun(N, OM) ->
+        %% tracing is a macro
+        case partisan_config:get(tracing, ?TRACING) of
+            true ->
+                lager:info("~p: Forwarding subscription for ~p to node: ~p", [node(), Node, N]);
+            false ->
+                ok
+            end,
+            OM ++ [{N, {protocol, {forward_subscription, Node}}}]
+            end, OutgoingMessages1, Membership0),
+    C = partisan_config:get(scamp_c, ?SCAMP_C_VALUE),
+    ForwardMessages = lists:map(fun(N) ->
+        case partisan_config:get(tracing, ?TRACING) of
+            true ->
+                lager:info("~p: Forwarding additional subs. for ~p to node: ~p", [node(), Node, N]);
+            false ->
+                ok
+        end,
+        {N, {protocol, {forward_subscription, Node}}}
+        end, select_random_sublist(state0, C)),
+    OutgoingMessages = OutgoingMessages2 ++ ForwardMessages,
+    State = State0#hiScamp{membership=Membership},
+    MembershipList = membership_list(State),
+    {ok, Membership, OutgoingMessages, State}.
+
+try_distance(#hiScamp{membership=Membership0, distance=Dist}=_State0, Node, Dict) ->
+    Dict = get(distance_metrics),
+    case dict:find(Node, Dict) == error of
+        true ->
+            erlang:error("Cannot attain node distain");
+        false ->
+            {ok, Dist} -> Dist
+    end,
+    Dist = element{2, Dist}.
+    
+
+distance(#hiScamp{membership=Membership0, heap=Heap, level1=L1, level2=L2, distance=Dist}=_State0, Node) ->
+    % A node j joins the system by sending a subscription request to the node s which is closest to it
+    Dict = dict:new(),
+    Threshold = 3,
+    Dist = try_distance(Node, Dict),
+    case Threshold > element{2, Dist} of
+        true -> 
+            % process this case using Scamp within this cluster
+            % subscription at level 1
+            %need to send copies out to other nodes -> gossip protocol
+            sets:add(Node, L1);
+        false ->
+                       % search through set to find another set
+                       % store two sets in the record
+            % subscription at level 2
+                       %input Cluster0
+            sets:add(Node, L2)
+    end,
+    heaps:add(Dist, Heap).
+
+get_centroid_two_clusters(#hiScamp{membership=Membership0, heap=Heap, level1=L1, level2=L2} =State0) ->
```

Comment:
I'm not sure why this is needed.  I don't see anything in the algorithm description that needs this?

Reply:
I'm sorry, but you're not sure about which part?

Reply:
This function.

Reply:
are you referring to get_centroid_two_clusters() or distance() because both of those functions are showing up in your comment

Reply:
```get_centroid_two_clusters()```

Reply:
No it doesn't say in the paper to do this, that was just how I thought about it.

Reply:
As far as I can tell from my read-through of the paper, the algorithm can operate without knowing this -- the protocol is designed for L2 to be maintained through just forwarding a la Scamp style.